### ISSUE TITLE
SEACAS CI tests failing due to push on 11/30/2017

### ISSUE THREAD

**Issue by bartlettroscoe** [trilinos/Trilinos#2039]
CC: @trilinos/seacas, @trilinos/framework, @gdsjaar, @bmpersc 

**Description:**

The push yesterday:

```
commit 57b14b2cbfa985f08090f635a298f6d963e18bd4
Author:     Brent Perschbacher <bmpersc@sandia.gov>
AuthorDate: Thu Nov 30 13:15:39 2017 -0700
Commit:     Brent Perschbacher <bmpersc@sandia.gov>
CommitDate: Thu Nov 30 16:02:19 2017 -0700

    Snapshot of sierra.base.git from commit ae935cf42a76fd89f31d4982cbcc013929ae3694
    
    From repository at sierra-git.sandia.gov:/git/sierra.base.git
    
    At commit:
    commit ae935cf42a76fd89f31d4982cbcc013929ae3694
    Author: Paul Wolfenbarger <prwolfe@sandia.gov>
    Date:   Thu Nov 30 11:17:59 2017 -0700
    
        seacas/Jamfile : remove "<define>DISABLE_PARALLEL_EPU"
    
        This will turn use of mpi in epu back on as the
        toolset and code changes have now been tested
        together and are working on both desktop and
        queued systems.

Commits pushed:
57b14b2 Snapshot of sierra.base.git from commit ae935cf42a76fd89f31d4982cbcc013929ae3694
ae3257c Snapshot of sierra.base.git from commit c58770384dc6fde65170c8d217488dd130e483dd
```

appears to have broken both of the CI builds at:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&display=project&filtercount=2&showfilters=1&filtercombine=and&field1=groupname&compare1=61&value1=Continuous&field2=buildstarttime&compare2=84&value2=now

The first CI iteration that failed is shown at:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=3248454

and the only commits pulled that CI iteration are from @bmpersc as shown at:

* https://testing.sandia.gov/cdash/viewNotes.php?buildid=3248455

It shows the failing tests:

| Name | Status | Time | Details |
| -------| ------- | ------ | ------ |
| SEACASExodus_exodus_unit_tests | Failed | 1s 640ms | Completed (Failed) |
| SEACASExodus_exodus_unit_tests_nc5_env | Failed | 1s 300ms | Completed (Failed) |

**Comment by bartlettroscoe** [comment]
I will go ahead and temporarily disable these two tests for the checkin-test-sems.sh script so it will not block people who are still using it.

**Comment by bartlettroscoe** [comment]
When I tried to disable and test and push with:

```
$ ./checkin-test-sems.sh --no-enable-fwd-packages --enable-all-packages=off \
  --enable-packages=SEACAS --do-all --push --rebase
```

I got two new failing SEACAS tests:

```
...
 5/16 Test #11: SEACASAprepro_aprepro_unit_test ...................***Failed  Required regular expression not found.Regex=[OVERALL FINAL RESULT: TEST PASSED .SEACASAprepro_aprepro_unit_test.
]  0.22 sec
...
15/16 Test #16: SEACASAprepro_aprepro_test_exodus .................***Failed  Required regular expression not 
88% tests passed, 2 tests failed out of 16
...

Label Time Summary:
SEACAS    =   7.65 sec (16 tests)

Total Test time (real) =   3.91 sec

The following tests FAILED:
	 11 - SEACASAprepro_aprepro_unit_test (Failed)
	 16 - SEACASAprepro_aprepro_test_exodus (Failed)
```

Therefore, my push was stopped.  I am running `make dashboard` in the `MPI_RELEASE_DEBUG_SHARED_PT` directory to post the results to CDash.

**Comment by gsjaardema** [comment]
I think that may be a probl;em with riunning the tests in parallel.  There
are some conflicts in filenames when ttwo or more tests are being run
simultaneously.

I do have this fixed in the SEACAS github, but I haven't synched it over to
Sierra and thus Trilinos yet.

..Greg

**Comment by bartlettroscoe** [comment]
> I do have this fixed in the SEACAS github, but I haven't synched it over to Sierra and thus Trilinos yet.

Then let's just disable these tests in Trilinos for now and then re-eanble them on next sync.  Any objections to doing that?

**Comment by bartlettroscoe** [comment]
I verified that if you run the SEACAS test suite with `ctest -j1,` the tests pass as shown at:

* https://testing.sandia.gov/cdash/viewTest.php?onlypassed&buildid=3250427

I will go ahead and disable these two tests as well in CI testing so that this does not trip anyone up using the checkin-test-sems.sh script and it will show the corresponding CI build clean.

**Comment by bartlettroscoe** [comment]
I just pushed the commit 73ecb117df6a2a2ce1e33ba9a982bbb0037192bc which should clean up the CI build shown at:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=66&value1=-MPI_RELEASE_DEBUG_SHARED_PT_CI&field2=groupname&compare2=61&value2=Continuous&field3=buildstarttime&compare3=84&value3=now

The other CI build and the other automated builds should still show these failures.  The @trilinos/framework team should decide how to address those failures.

**Comment by bartlettroscoe** [comment]
After my push, the CI build is now clean as shown at:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=3250450

But note that the other CI build is still going to show these failures and well as all of the other nightly builds of SEACAS as you can see at:

* https://testing.sandia.gov/cdash/index.php?project=Trilinos&subproject=SEACAS&date=2017-12-02

When the next sync of SEACAS is done, the commit 73ecb11 should be reverted.

**Comment by bmpersc** [comment]
@bartlettroscoe,
  @gdsjaar is right that these are tests that have a race condition exposed by ctest running in parallel. I was aware of and had been talking with Greg about these. When I ran the tests last week and they passed I did check to see if there was a fix that looked like it addressed this issue and saw something that looked like it did. However, looking at it again I was mistaken. I didn't notice that it was only for Aprepro and not a fix for these tests.

If I can't get an actual fix in before the end of today I will revert the offending commit.

**Comment by rppawlo** [comment]
The Nov 30th seems to have broken exodus output for EMPIRE as well.

@bathmatt

**Comment by bartlettroscoe** [comment]
> The Nov 30th seems to have broken exodus output for EMPIRE as well.

The problem was just supposed to be with the tests (race conditions) and not the library code itself.

I think it is getting to be about time to pick back up #1751 so that we can let @gsjaardema snapshot SEACAS into SIERRA and Trilinos so that he can directly work with the various code teams to address issues ASAP.  We just need to write down a few process things and make sure that @gsjaardema and @bmpersc communicate so that they don't step on each others toes.

**Comment by bathmatt** [comment]
I have confirmed that this broke EMPIRE< however, not sure if it is an old standing bug which got tickled or something new.

**Comment by bmpersc** [comment]
@bartlettroscoe based on the discussion going on between @bathmatt and @gsjaardema  in email it seems like there is a good chance @bathmatt's issue is actually in Panzer and was exposed by some fixes made to STK with respect to flushing output. 

So the question I have is do we back out changes to package A because it exposed a bug in package B or do we expect package B to fix its bugs?

FWIW I am currently testing the patch Greg had for fixing the testing race condition in SEACAS which assuming it passes all tests I was considering pushing that as opposed to backing out the changes made. Clearly this won't fix @bathmatt's issue, but it is unclear if the issue is in these changes.

**Comment by bartlettroscoe** [comment]
@bmpersc,

> So the question I have is do we back out changes to package A because it exposed a bug in package B or do we expect package B to fix its bugs?

That is a good question.  For packages is Trilinos, if changes to Package A expose defects in Package B that cause automated tests to fail for the standard CI builds or any of the builds associated with the automated PR testing (i.e. #1155) then I think it is clear that you back our the changes ASAP.  But other than that, there is not a clear answer.  I think we need to address these others situations on a case-by-case basis.  How bad is the situation for Panzer and EMPIRE?  Is there an easy workaround for now?  Or, is it a large effort to address the issue?  Answers to these would likely address the situation.

But note that once we have the proper git.git-like workflow setup for EMPIRE, then changes like this should not directly impact Drekar and EMPIRE.  In fact, for that reason perhaps we need to pull the trigger on that workflow, even if all of the documentation is not completed yet.  I will comment on this in https://gitlab.sandia.gov/EM-Plasma/EMPIRE/issues/162.

**Comment by bathmatt** [comment]
I'm not sure this is a bug in panzer, not sure where the bug is honestly but thinking actually STK, but I have a hack work around,

**Comment by bartlettroscoe** [comment]
@bmpersc,

> When I ran the tests last week and they passed I did check to see if there was a fix that looked like it addressed this issue and saw something that looked like it did.

Did the automated PR testing not catch this?  (The checkin-test-sems.sh script testing as well as the post-push CI servers caught a few of these test failures.)

**Comment by bmpersc** [comment]
I have reverted both the seacas and stk update commits from 11/30/2017. This should address this seacas test race condition as well as Matt's issue.

@bathmatt If you believe that this is an issue with STK alone you need to file a ticket with them about the issues you were seeing so they can address it before the next integration. Especially since we aren't testing your project before we integrate it will be hard for us to know if it is fixed without risking breaking you again.

@bartlettroscoe I did not run the PR testing because the integration predates the PR testing. We did decide during the framework standup to make the integration use the PR testing in addition to its own testing. The testing that the integration runs ran into the race condition too, in fact it was the integration testing where it was first uncovered and I reported it to Greg. However, since it is a race condition on which test writes to a file first it is a fluke that any testing caught it.

**Comment by bartlettroscoe** [comment]
> @bathmatt If you believe that this is an issue with STK alone you need to file a ticket with them about the issues you were seeing so they can address it before the next integration. Especially since we aren't testing your project before we integrate it will be hard for us to know if it is fixed without risking breaking you again.

@bathmatt, in case you do want to test the SIERRA version of STK, all you need to do is to clone the Sierra.base repo, put a symlink to the `stk` source dir of that repo under the local Trilinos base dir and the configure Trilinos as you would but add `-DSTK_SOURCE_DIR_OVERRIDE:STRING=stk` (you might also need to do this for SEACAS as well).  The will use that version of STK and ignore the version under Trilinos/packages/stk.  That way you (or some other Drekar or EMPIRE developer) can interact with the STK developers much more directly.   I can give more details and exact instructions for anyone who is interested in this type of workflow.

**Comment by bmpersc** [comment]
after reverting the integration commits the continuous builds are back to normal so I'm closing this. The other issues discussed here can be moved to more appropriate issues.
